### DISCUSSION TITLE
Figure width is not respected when using CairoMakie

### DISCUSSION THREAD

**Discussion by tpoisot** [quarto-dev/quarto-cli#10602] [Q&A]
### Description

I am running into a problem with quarto 1.5.47 when generating a pdf from a file using CairoMakie to generate the figures. Specifically, there seems to be no way to get the figures in the PDF to use the textwidth.

This used to work on previous versions of quarto, and this is not an issue with e.g. pandoc, for the record.

Here is an example document that shows the issue: the scatter plots overflow from the text.

~~~
---
title: "Untitled"
format: pdf
engine: julia
---

```{julia}
import Pkg
Pkg.activate(joinpath(homedir(), ".julia", "environments", "v1.10"))
```

```{julia}
using CairoMakie
```

Plot with no options

```{julia}
scatter(rand(1000))
```

Plot with size options

```{julia}
#| fig-width: 100%
scatter(rand(1000))
```
~~~

**Comment by cscheid** [comment]
In Jupyter, we currently only support `fig-width` etc as a document-wide option.

**Reply by tpoisot** [reply]
Sure, but even this is not working. Maybe this is because I'm using `engine: julia`.

Regardless -- is there a way to ensure that in the PDF, figures do not get larger than `\textwidth`?

**Comment by tpoisot** [comment]
I've been trying to troubleshoot this a little bit more, and [the documentation](https://quarto.org/docs/authoring/figures.html) says

> By default figures are displayed using their actual size (subject to the width constraints imposed by the page they are rendered within). You can change the display size by adding the width and height attributes to the figure.

This refers, I suppose, to the fact that my `.tex` output will get something like:

~~~
\includegraphics[width=20.83333in,height=20.83333in]{chapters/classification_files/figure-pdf/fig-classification-decision-output-1.png}
~~~

which *should* not overflow from the page because the template includes

~~~
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
~~~

And yet - the figures *do* overflow:

![image](https://github.com/user-attachments/assets/996d02ad-cba6-4c2d-afde-870c6a1fa26d)

**Reply by cscheid** [reply]
> `\includegraphics[width=20.83333in,height=20.83333in]`

This feels unlikely to be on purpose (20 inches on a ~8.5x11 page?) Can we get a minimal repro of the problem?

**Reply by tpoisot** [reply]
Definitely not on purpose (and also the root of the problem!), and the reason why is: quarto decides on the physical size of the figure by doing pixels * 96 (assuming 96 PPI). So for high-resolution figures, this creates this exact problem. 

I'll make a reproducible example when I'm back home, but that's the behavior causing the issue.

**Reply by tpoisot** [reply]
Here is an example file

~~~
---
title: "Untitled"
format:
    pdf:
        keep-tex: true
engine: julia
---

```{julia}
import Pkg
Pkg.activate(joinpath(homedir(), ".julia", "environments", "v1.10"))
```

```{julia}
plotsize = (600, 300)
scaling = 3.14159265
dpi = 96
```

```{julia}
using CairoMakie
CairoMakie.activate!(; px_per_unit = scaling, type="png")
```

The physical size of the figure in inches will be

```{julia}
(plotsize .* scaling) ./ dpi
```

```{julia}
f = Figure(; size=plotsize)
ax = Axis(f[1,1])
scatter!(ax, rand(100))
f
```
~~~

The relevant part of the tex file is

~~~tex

\includegraphics[width=19.63542in,height=9.8125in]{test_files/figure-pdf/cell-6-output-1.png}

~~~

And indeed, the PDF compiled with quarto reads:

~~~
The physical size of the figure in inches will be (19.6349540625, 9.81747703125)
~~~

And the figure actually overflows:

![image](https://github.com/user-attachments/assets/340224c6-811e-4e71-a556-ab2fa1f12569)

So when [the documentation](https://quarto.org/docs/authoring/figures.html) says

> By default figures are displayed using their actual size (subject to the width constraints imposed by the page they are rendered within).

The first part of the sentence is true, *but* the second part is not: the width of the page in which the figure is rendered is not respected, and the figure is larger than `\textwidth` / taller than `\textheight`.

The next sentence in the documentation is

> You can change the display size by adding the width and height attributes to the figure.

But adding `#| width: 5in` to the code that generates the figure has no effect / isn't reflected in the tex file.

**Reply by cscheid** [reply]
> > You can change the display size by adding the width and height attributes to the figure.
> 
> But adding `#| width: 5in` to the code that generates the figure has no effect / isn't reflected in the tex file.

It's understandable that you'd take the former to mean the latter, but that part of the documentation is talking about adding attributes to _figures_, and not to code cell "adding the width and height attributes to the figure".

This is additionally awkward because adding width to an _R_ cell works, but that's a knitr-specific feature, and there's currently no clean way to make this work in other languages.

**Reply by cscheid** [reply]
Here's a relevant part of the `keep-md: true` output:

````
::: {.cell execution_count=1}
``` {.julia .cell-code}
f = Figure(; size=plotsize)
ax = Axis(f[1,1])
scatter!(ax, rand(100))
f
```

::: {.cell-output .cell-output-display execution_count=1}
![](issue-10602_files/figure-pdf/cell-6-output-1.png){width=1885 height=942 fig-pos='H'}
:::
:::
````

This is the source of the problem, and there's not much that Quarto can do here. Those width and height values (`width=1885 height=942`) are going to be interpreted in inches, and so that's going to cause trouble.

The next step is to understand why those values are being provided like that, but I'm not a CairoMakie expert (or even user), so I can't help much.

@jkrumbiegel, sorry to tag you here, but maybe you have some input on what's going on here?

**Reply by tpoisot** [reply]
A workaround is to specify a higher DPI and a value of `px_per_unit` in CairoMakie, then to scale figures very explicitely by knowing their physical dimension. For example, a 7in by 6in figure with `dpi: 600` and `px_per_unit=5` needs to have the following dimensions:

~~~julia
f = Figure(; size=(7, 6).*(600/5))
~~~

For the specific document in which I'm having this problem, this works since I know the exact dimension of the textbox.

**Reply by jkrumbiegel** [reply]
So I've run a little test. This is the notebook I ran, once with `fig-dpi: 96` and once with `fig-dpi: 192`:

````
---
title: "Untitled"
format: pdf
engine: julia
keep-md: true
fig-dpi: 96
execute: 
  daemon: false
---

```{julia}
using CairoMakie

function testfig(size, px_per_unit)
    f = Figure(; size, backgroundcolor = :tomato)
    Label(f[1, 1], "$size\npx_per_unit=$px_per_unit", fontsize = 20)
    f
end

testfig((200, 100), Makie.current_default_theme()[:CairoMakie][:px_per_unit][])
```

```{julia}
testfig((300, 200), Makie.current_default_theme()[:CairoMakie][:px_per_unit][])
```
````

This is the result for 96 dpi, the `px_per_unit` value is set to 1.0 (this is done by a CairoMakie package hook within QuartoNotebookRunner which picks up `fig-dpi`):

<img width="784" alt="image" src="https://github.com/user-attachments/assets/0ba1c289-ab9c-4dd2-80ee-5cdbdc9b5ba2">

and in the intermediate md, I get these sizes annotated for the images:

```
width=200 height=100
width=300 height=200
```

This is the result for 192 dpi, the `px_per_unit` value is set to 2.0 instead:

<img width="572" alt="image" src="https://github.com/user-attachments/assets/b6977f21-6e21-4006-85b3-59a543595ab6">

and in the intermediate md, I get these sizes instead:

```
width=400 height=200
width=600 height=400
```

So what can we conclude from that?

- CairoMakie picks up the `fig-dpi` and sets `px_per_unit` to `figdpi / 96`. This is correct because one "unit" in CairoMakie is a CSS pixel and there are 96 of these in an inch. So for 96 dpi or pixels per inch we need a px/unit ratio of 1.
- The pixel sizes of the rendered pngs correspond correctly to that, a 300x200 units figure has resolution 300x200 pixels when rendered with 96 dpi and 600x400 pixels when rendered with 192 dpi.
- The annotation in the intermediate md doesn't make sense, this just mirrors the pixel size of the png, but not the logical size of the image (which would correspond to the `Figure(size = ...)`). Because of this, images with higher dpi get blown up in the resulting latex doc. The two images in doc 1 should be the same resulting size as in doc 2.

If I render only an ipynb with `QuartoNotebookRunner` directly, I find these pieces of metadata in there:

```
"data": {
                        "image/png": some image data
                    "metadata": {
                        "image/png": {
                            "width": 300,
                            "height": 200
                        }
                    }
```

These width/height things just report the pixel resolution of the png image. I assume that this is where the problem originates, although I'm not sure if quarto uses this value incorrectly as a logical size or QuartoNotebookRunner should populate this differently, as a logical size and not a resolution.

**Reply by cscheid** [reply]
This is super helpful, @jkrumbiegel, thank you! One small comment:

> If I render only an ipynb with `QuartoNotebookRunner` directly

Quarto's `jupyter` engine detects and handles a `keep-ipynb: true` option, and produces a `document_name.quarto_ipynb` file, which holds the .ipynb result of executing the code before converting it to Markdown for further processing. I wonder if it makes sense for the `julia` engine to do so as well.

**Reply by jkrumbiegel** [reply]
Mentioning here that I'm planning to have QuartoNotebookRunner write a dpi-corrected size into this metadata field if dpi is available in the png file https://github.com/PumasAI/QuartoNotebookRunner.jl/pull/248

If not, images will continue to be sized as if they had 96 dpi, as it currently works.

**Comment by nguiard** [comment]
Hi,
I'm running into this issue. While setting fig-dpi to 96 works to have the correct size, when I try actually plotting something, the resulting image in the pdf is so blurry that it is completely unfit for publication.

Code:

``````qmd
---
engine: julia
fig-dpi: 96
format: pdf
---

Some text with default size for reference.

```{julia}
using CairoMakie

function draw()
    fig1 = Figure(; size=(600, 400))
    ax1 = Axis(fig1[1, 1])
    lines!(ax1, 1:10, sin.(1:10))

    fig1
end

draw()
```
``````

While the result might look fine from afar:

![Screenshot from 2025-04-05 14-00-23](https://github.com/user-attachments/assets/fcedebe6-5322-4e9d-aed1-61fd8d1ffaee)

If you zoom in just a little bit, things get very blurry:

![Screenshot from 2025-04-05 14-00-43](https://github.com/user-attachments/assets/404779dd-42f8-4624-8863-2a252f8248db)

Is there a way I can make this work? Maybe I've missed something?

**Reply by nguiard** [reply]
Oh actually, for anyone wondering: just setting CairoMakie to output SVG by adding `CairoMakie.activate!(type = "svg")` at the start of the function does the trick. I can then remove the fig-dpi setting.

**Reply by jkrumbiegel** [reply]
This is all fixed on prerelease quarto and newer Makies. Then the dpi will be respected but the size will also be correct. Svg doesn't make sense for some plots, when you have a lot of data points in a scatter, for example.

**Reply by nguiard** [reply]
Very good to know! Thanks

**Reply by nguiard** [reply]
Hum, I just tried on prerelease 1.7.21, with CairoMakie 0.13.2, and the problem still appears for me... Am I missing something?

To be sure, I tried running your test code (further up in this issue) in a clean context: 

``````
---
title: "Untitled"
format: pdf
engine: julia
keep-md: true
fig-dpi: 96
execute: 
  daemon: false
---

```{julia}
using CairoMakie

function testfig(size, px_per_unit)
    f = Figure(; size, backgroundcolor = :tomato)
    Label(f[1, 1], "$size\npx_per_unit=$px_per_unit", fontsize = 20)
    f
end

testfig((200, 100), Makie.current_default_theme()[:CairoMakie][:px_per_unit][])
```

```{julia}
testfig((300, 200), Makie.current_default_theme()[:CairoMakie][:px_per_unit][])
```
``````

And things behave the same as before. Changing `px_per_unit` has no effect other than being correctly written in the colored boxes, and changing `fig-dpi` changes the overall size.

**Reply by nguiard** [reply]
Oooh. For some reason, QuartoNotebookRunner didn't update at once...

And at some point experimenting and rerunning quarto, I saw quarto update, and in particular `QuartoNotebookRunner v0.11.6 ⇒ v0.15.0`. And yeah, that fixed everything... So, sorry for the false warning.

Not sure what triggered the update though, I just ran `quarto preview` maybe for the 10th time in different contexts trying things, and at some point it ran an update of its environment. Is there a way to trigger this manually btw?

**Reply by jkrumbiegel** [reply]
If you had a new CairoMakie and a prerelease quarto you probably still had the older QuartoNotebookRunner server process running? I think it's the server that reads out the dpi data so if you had an older one that would explain that there was no change. You can try `quarto call engine julia status` and `quarto call engine julia stop` or `quarto call engine julia kill`

Edit: Ah your message loaded only after I had written mine. QuartoNotebookRunner will also only update when the control server first starts up.
